iquip/apps/dataviewer.py

```diff
@@ -634,24 +733,213 @@ def datasetName(self) -> str:
+            self.frame.datasetName(),
+            self.constants.proxy_ip,  # pylint: disable=no-member
+            self.constants.proxy_port,  # pylint: disable=no-member
+            self.setDataset,
+        )
+        self.thread.start()
+
+    @pyqtSlot(np.ndarray, list, list)
+    def setDataset(
+        self,
+        dataset: np.ndarray,
+        parameters: List[str],
+        units: List[Optional[str]],
+    ):
+        """Sets the dataset to show and updates the axis viewer.
+        
+        Args:
+            See SimpleScanDataPolicy.
+        """
+        self.policy = SimpleScanDataPolicy(dataset, parameters, units)
+        self.frame.sourceWidget.setParameters(parameters, units)
+
+    @pyqtSlot(tuple)
+    def setAxis(self, axis: Sequence[int]):
+        """Given the axis information, draws the main plot.
+        
+        Args:
+            axis: See updateMainPlot().
+        """
+        self.axis = axis
+        dataType = self.frame.dataPointWidget.dataType()
+        self.updateMainPlot(axis, dataType)
+
+    @pyqtSlot(DataPointWidget.DataType)
+    def setDataType(self, dataType: DataPointWidget.DataType):
+        """Given the data type, draws the main plot.
+        
+        Args:
+            dataType: See updateMainPlot().
+        """
+        self.updateMainPlot(self.axis, dataType)
+
+    def updateMainPlot(self, axis: Sequence[int], dataType: DataPointWidget.DataType):
+        """Updates the main plot.
+        
+        Args:
+            axis: See SimpleScanDataPolicy.extract().
+            dataType: Target data type.
+        """
+        if self.policy is None:
+            return
+        reduce = self._reduceFunction(dataType)
+        data, axes = self.policy.extract(axis, reduce)
+        self.frame.mainPlotWidget.setData(data, axes)
+        index = self.dataPointIndex
+        if data.ndim == len(index) and np.all(np.less(index, data.shape)):
+            self.selectDataPoint(index)
+        else:
+            self.selectDataPoint((0,) * data.ndim)
+
+    @pyqtSlot(tuple)
+    def selectDataPoint(self, index: Tuple[int, ...]):
+        """Selects a data point at the given index.
+        
+        Args:
+            index: The index of the target data point, in the dataset array.
+        """
+        if self.policy is None:
+            return
+        self.dataPointIndex = index
+        _, symbols = self.policy.symbolize(self.axis)
+        data_indices = np.all(symbols.T == index, axis=1)
+        data = self.policy.dataset[:, 0][data_indices].astype(int)
+        for dataType in DataPointWidget.DataType:
+            value = self._reduceFunction(dataType)(data)
+            self.frame.dataPointWidget.setValue(value, dataType)
+        bins, counts = np.unique(data, return_counts=True)
+        self.frame.dataPointWidget.setHistogramData(bins, counts)
+
+    @pyqtSlot(int)
+    def setThreshold(self, threshold: int):  # pylint: disable=unused-argument
```

Comment:
How about naming `threshold` with a leading underscope in order to bypass the pylint warning?

Reply:
Considering inheritance and overriding, I think we should not name the arguments like that (put an underscore to avoid pylint warning), because it will cause trouble when we call overridden methods with keyword arguments.
(Even if `DataViewerApp` might not be inherited, for consistency!)

For example:
```python
class A:
    def f(self, x, _y):
        """Uses x but not _y."""

class B(A):
    def f(self, _x, y):
        """Overridden, uses y but not _x."""

def test(obj: A):  # obj can be an instance of A or B
    obj.f(x=0, y=0)  # error, one can never call f() with keyword arguments
```

Anyway, I just removed the `threshold` argument since it is okay for slots to have less arguments than the connected signals.

Reply:
Oh, I see!

BTW, your solution looks so cool!